### ISSUE TITLE
Allow editorconfig Naming Rules/Styles/SymbolSpecs to have uppercase letters

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#40705]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/869465/editorconfig-naming-convention-not-working-c-visua.html)._

---
I am trying to use .editorconfig at solution level to define naming conventions for the two C# projects in it. I have latest Visual Studio as of date, updated just today to 16.4.2. It is giving warnings where not expected. For example, for class names, property names, enum names it says that it should begin with lower case, whereas as per my convention in editorconfig, they are expected in pascal case. Similarly for private fields in class, I have defined field name to begin with underscore ("_") character but it is reporting that "_" is unexpected.

I am sure that it is using my editorconfig file because I see changes in messages if I deliberately change some parts just to see whether there is any impact. However, result is not as per expectation.
Some lines where error is reported but not expected are:
public partial class WebApplication : AnApplication - "The first word, 'WebApplication', must begin with a lower case character."
private readonly string _driverPath; - "Prefix '_' is not expected"
public RemoteWebDriver WebDriver { get; private set; } -  "The first word, 'WebDriver', must begin with a lower case character."

Are there more settings that need to be done? Is there an error in my editorconfig inputs?

Here is the relevant content of my editorconfig file.

```ini
# Naming rules definitions

dotnet_naming_rule.interfacesPascalWithI.symbols = interfaces
dotnet_naming_rule.interfacesPascalWithI.style = IP
dotnet_naming_rule.interfacesPascalWithI.severity = suggestion

dotnet_naming_rule.publicMethodsPascal.symbols = pubMethods
dotnet_naming_rule.publicMethodsPascal.style = P
dotnet_naming_rule.publicMethodsPascal.severity = suggestion

dotnet_naming_rule.defaultMethodsCamel.symbols = methods
dotnet_naming_rule.defaultMethodsCamel.style = c
dotnet_naming_rule.defaultMethodsCamel.severity = suggestion

dotnet_naming_rule.staticFieldsPascal.symbols = staticFields
dotnet_naming_rule.staticFieldsPascal.style = c
dotnet_naming_rule.staticFieldsPascal.severity = warning

dotnet_naming_rule.constFieldsPascal.symbols = constFields
dotnet_naming_rule.constFieldsPascal.style = c
dotnet_naming_rule.constFieldsPascal.severity = warning

dotnet_naming_rule.privateFieldsUnderscore.symbols = prvFields
dotnet_naming_rule.privateFieldsUnderscore.style = _c
dotnet_naming_rule.privateFieldsUnderscore.severity = suggestion

dotnet_naming_rule.defaultFieldsCamel.symbols = fields
dotnet_naming_rule.defaultFieldsCamel.style = c
dotnet_naming_rule.defaultFieldsCamel.severity = suggestion

dotnet_naming_rule.defaultPascal.symbols = default
dotnet_naming_rule.defaultPascal.style = P
dotnet_naming_rule.defaultPascal.severity = suggestion

# Style Definitions
dotnet_naming_style. P.capitalization = pascal_case

dotnet_naming_style. IP.capitalization = pascal_case
dotnet_naming_style. IP.required_prefix = I

dotnet_naming_style.c.capitalization = camel_case

dotnet_naming_style._c.capitalization = camel_case
dotnet_naming_style._c.required_prefix = _

# Symbols Definitions
dotnet_naming_symbols.default.applicable_kinds = *
dotnet_naming_symbols.default.applicable_accessibilities = *

dotnet_naming_symbols.interfaces.applicable_kinds = interface
dotnet_naming_symbols.interfaces.applicable_accessibilities = *

dotnet_naming_symbols.pubMethods.applicable_kinds = method
dotnet_naming_symbols.pubMethods.. applicable_accessibilities = public

dotnet_naming_symbols.methods.applicable_kinds = method
dotnet_naming_symbols.methods.applicable_accessibilities = *

dotnet_naming_symbols.staticFields.applicable_kinds = field
dotnet_naming_symbols.staticFields.applicable_accessibilities = *
dotnet_naming_symbols.staticFields.required_modifiers = static

dotnet_naming_symbols.constFields.applicable_kinds = field
dotnet_naming_symbols.constFields.applicable_accessibilities = *
dotnet_naming_symbols.constFields.required_modifiers = const

dotnet_naming_symbols.prvFields.applicable_kinds = field
dotnet_naming_symbols.prvFields.applicable_accessibilities = private

dotnet_naming_symbols.fields.applicable_kinds = field
dotnet_naming_symbols.fields.applicable_accessibilities = *
```

---
#### Visual Studio Feedback System on 12/30/2019, 01:49 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 12/30/2019, 02:21 PM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

**Comment by dpoeschl** [comment]
The original issue here can be worked around by lowercasing the names of rules, styles, and symbols. One quick way to do this in Visual Studio is to select the entire block, press Ctrl+u (this will lowercase everything in the selection), and then go back and re-capitalize the prefixes where appropriate.

**Comment by dpoeschl** [comment]
@jmarolf Do you remember if this was intentional for some reason, or fallout of the way .editorconfigs are initially parsed by the editor?

**Comment by jmarolf** [comment]
no idea if this is the issue, but this section

```ini
dotnet_naming_style. P.capitalization = pascal_case

dotnet_naming_style. IP.capitalization = pascal_case
dotnet_naming_style. IP.required_prefix = I
```

Has a space before `P` and `IP`. This would cause things to go wrong. If that is just a formatting error on githubs side then I would assume that this is a roslyn bug and not an editorconfig parser bug as I see nothing else wrong with the users editorconfig and this scenario should work

**Comment by dpoeschl** [comment]
`but this section has a space before P and IP`

Even with that fixed, I needed to make everything lowercase.

**Comment by jmarolf** [comment]
Hmmm, now I am curious

so [this](https://dev.azure.com/devdiv/DevDiv/_git/CodingConventions?path=%2Fsrc%2FMicrosoft.VisualStudio.CodingConventions%2FEditorConfig%2FEditorConfigParser.cs&version=GBmaster&line=17&lineEnd=18&lineStartColumn=1&lineEndColumn=1&lineStyle=plain) is the regex that the vs parser uses.

```regex
^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$
```


running this example program on [try.dot.net](https://try.dot.net/)
```csharp
using System.Text.RegularExpressions;
using static System.Console;

class Program {
    static void Main(string[] args) {
        foreach (var line in lines) {
            var propMatches = _propertyMatcher.Matches(line);
            if (propMatches.Count > 0 && propMatches[0].Groups.Count > 0) {
                WriteLine("Group captured");
            }
            else {
                WriteLine($"Group '{line}' not captured");
            }
        }
    }

    static Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);

    static string[] lines = new[] {
        "dotnet_naming_style. P.capitalization = pascal_case",
        "dotnet_naming_style. IP.capitalization = pascal_case",
        "dotnet_naming_style. IP.required_prefix = I",
    };
}
```

yields these results
```
Group 'dotnet_naming_style. P.capitalization = pascal_case' not captured
Group 'dotnet_naming_style. IP.capitalization = pascal_case' not captured
Group 'dotnet_naming_style. IP.required_prefix = I' not captured
```

suggesting that these options were not captured at all before.

`Dictionary<string, string>` is the type used throughout to store options and the default comparer for `Dictionary<string, string>` is `GenericEqualityComparer<string>` which is case-sensitive.

This would suggest that the problem lies somewhere in [EditorConfigNamingStyleParser.cs](https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs). While we did introduce some logic to resolve duplicate titles in https://github.com/dotnet/roslyn/pull/36366 I don't think that would cause this.

**Comment by dpoeschl** [comment]
To simplify a bit, here's a single case that doesn't work in the current implementation:

```editorconfig
root = true

[*.{cs,vb}]

# Naming rules definitions
dotnet_naming_rule.interfacesPascalWithI.symbols = interfaces
dotnet_naming_rule.interfacesPascalWithI.style = IP
dotnet_naming_rule.interfacesPascalWithI.severity = suggestion

# Style Definitions
dotnet_naming_style.IP.capitalization = pascal_case
dotnet_naming_style.IP.required_prefix = I

# Symbols Definitions
dotnet_naming_symbols.interfaces.applicable_kinds = interface
dotnet_naming_symbols.interfaces.applicable_accessibilities = *
```

This doesn't work because the raw options dictionary we are given has all keys lowercased, but the values correctly cased. So when we try to look up "IP", we can't find it because the keys for it have been lowercased.

**Comment by jmarolf** [comment]
hmm, it looks like lowering the casing is not done in [EditorConfigParser.cs](https://dev.azure.com/devdiv/DevDiv/_git/CodingConventions?path=%2Fsrc%2FMicrosoft.VisualStudio.CodingConventions%2FEditorConfig%2FEditorConfigParser.cs&version=GBmaster) for VS but is done in the compiler implementation

https://github.com/dotnet/roslyn/blob/3892c636a33f7c3f2e27bc596ec6c43168b57854/src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs#L119-L124

specifically here:

https://github.com/dotnet/roslyn/blob/3892c636a33f7c3f2e27bc596ec6c43168b57854/src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs#L167-L171

I think we are going to want to change one implementation or the other

CC: @agocke
